diff --git a/cc/containers/sdk/BUILD b/cc/containers/sdk/BUILD
index c7565ee98..7eb2de0c3 100644
--- cc/containers/sdk/BUILD
+++ cc/containers/sdk/BUILD
@@ -15,7 +15,7 @@
 #
 
 package(
-    default_visibility = ["//:default_visibility"],
+    default_visibility = ["//visibility:public"],
     licenses = ["notice"],
 )
 
diff --git a/oak_attestation_verification/BUILD b/oak_attestation_verification/BUILD
index e355a02dd..673ff3c7c 100644
--- oak_attestation_verification/BUILD
+++ oak_attestation_verification/BUILD
@@ -17,7 +17,7 @@
 load("@rules_rust//rust:defs.bzl", "rust_library", "rust_test_suite")
 
 package(
-    default_visibility = ["//:default_visibility"],
+    default_visibility = ["//visibility:public"],
     licenses = ["notice"],
 )
 
@@ -74,6 +74,7 @@ rust_library(
     srcs = glob(["src/**"]),
     compile_data = ["//oak_attestation_verification/data:amd_ask_certs"],
     crate_features = ["regex"],
+    crate_name = "oak_attestation_verification",
     deps = [
         "//oak_dice",
         "//oak_proto_rust",
diff --git a/oak_client/BUILD b/oak_client/BUILD
index b335325d0..7cc01d7f0 100644
--- oak_client/BUILD
+++ oak_client/BUILD
@@ -17,7 +17,7 @@
 load("@rules_rust//rust:defs.bzl", "rust_library")
 
 package(
-    default_visibility = ["//:internal"],
+    default_visibility = ["//visibility:public"],
     licenses = ["notice"],
 )
 
diff --git a/oak_containers/agent/BUILD b/oak_containers/agent/BUILD
index 94c2ce0ba..285185713 100644
--- oak_containers/agent/BUILD
+++ oak_containers/agent/BUILD
@@ -17,7 +17,7 @@
 load("@rules_rust//rust:defs.bzl", "rust_binary", "rust_library")
 
 package(
-    default_visibility = ["//:internal"],
+    default_visibility = ["//visibility:public"],
     licenses = ["notice"],
 )
 
diff --git a/oak_containers/launcher/BUILD b/oak_containers/launcher/BUILD
index 9dd3bf07c..d89e60c0d 100644
--- oak_containers/launcher/BUILD
+++ oak_containers/launcher/BUILD
@@ -17,7 +17,7 @@
 load("@rules_rust//rust:defs.bzl", "rust_binary", "rust_library")
 
 package(
-    default_visibility = ["//:internal"],
+    default_visibility = ["//visibility:public"],
     licenses = ["notice"],
 )
 
diff --git a/oak_containers/orchestrator/BUILD b/oak_containers/orchestrator/BUILD
index 0ae89dab8..8b42d6248 100644
--- oak_containers/orchestrator/BUILD
+++ oak_containers/orchestrator/BUILD
@@ -17,7 +17,7 @@
 load("@rules_rust//rust:defs.bzl", "rust_binary", "rust_library", "rust_test")
 
 package(
-    default_visibility = ["//:internal"],
+    default_visibility = ["//visibility:public"],
     licenses = ["notice"],
 )
 
diff --git a/oak_crypto/BUILD b/oak_crypto/BUILD
index 8616d82b1..eb5b4ad1d 100644
--- oak_crypto/BUILD
+++ oak_crypto/BUILD
@@ -18,7 +18,7 @@ load("@rules_rust//rust:defs.bzl", "rust_library", "rust_test")
 load("//bazel:defs.bzl", "either_platform")
 
 package(
-    default_visibility = ["//:default_visibility"],
+    default_visibility = ["//visibility:public"],
     licenses = ["notice"],
 )
 
diff --git a/oak_proto_rust/BUILD b/oak_proto_rust/BUILD
index 47721ccdc..933ed2d3d 100644
--- oak_proto_rust/BUILD
+++ oak_proto_rust/BUILD
@@ -20,7 +20,7 @@ load("@rules_rust//rust:defs.bzl", "rust_library")
 load("//bazel:defs.bzl", "either_platform")
 
 package(
-    default_visibility = ["//:internal"],
+    default_visibility = ["//visibility:public"],
     licenses = ["notice"],
 )
 
diff --git a/oak_proto_rust/grpc/BUILD b/oak_proto_rust/grpc/BUILD
index a4089b218..4942e5fb4 100644
--- oak_proto_rust/grpc/BUILD
+++ oak_proto_rust/grpc/BUILD
@@ -18,7 +18,7 @@ load("@rules_rust//cargo:defs.bzl", "cargo_build_script")
 load("@rules_rust//rust:defs.bzl", "rust_library")
 
 package(
-    default_visibility = ["//:internal"],
+    default_visibility = ["//visibility:public"],
     licenses = ["notice"],
 )
 
diff --git a/proto/BUILD b/proto/BUILD
index 36884bcbf..85185d66f 100644
--- proto/BUILD
+++ proto/BUILD
@@ -20,7 +20,7 @@ load("@rules_java//java:defs.bzl", "java_proto_library")
 load("@rules_proto//proto:defs.bzl", "proto_library")
 
 package(
-    default_visibility = ["//:default_visibility"],
+    default_visibility = ["//visibility:public"],
     licenses = ["notice"],
 )
 
diff --git a/proto/attestation/BUILD b/proto/attestation/BUILD
index a216260f0..01ea01d9b 100644
--- proto/attestation/BUILD
+++ proto/attestation/BUILD
@@ -17,7 +17,7 @@
 load("@bazel_skylib//rules:build_test.bzl", "build_test")
 
 package(
-    default_visibility = ["//:default_visibility"],
+    default_visibility = ["//visibility:public"],
     licenses = ["notice"],
 )
 
diff --git a/proto/containers/BUILD b/proto/containers/BUILD
index 237d869b1..1cd6e810a 100644
--- proto/containers/BUILD
+++ proto/containers/BUILD
@@ -20,7 +20,7 @@ load("@rules_cc//cc:defs.bzl", "cc_proto_library")
 load("@rules_proto//proto:defs.bzl", "proto_library")
 
 package(
-    default_visibility = ["//:default_visibility"],
+    default_visibility = ["//visibility:public"],
     licenses = ["notice"],
 )
 
