diff --git a/oak_containers_launcher/src/qemu.rs b/oak_containers_launcher/src/qemu.rs
index 88472aa3bb..fa99a7ecb2 100644
--- oak_containers_launcher/src/qemu.rs
+++ oak_containers_launcher/src/qemu.rs
@@ -92,6 +92,14 @@ pub struct Params {
     /// SevSnp, or Tdx (TDX is unimplemented yet)
     #[arg(long, required = false, value_enum, default_value_t = VmType::Default)]
     pub vm_type: VmType,
+
+    /// Forward message from VM to launcher running on host.
+    #[arg(long, required = false, default_value_t = 8889)]
+    pub hats_launcher_port: u16,
+
+    /// Forward messages from host to VM.
+    #[arg(long, required = false, default_value_t = 50051)]
+    pub host_forwarding_port: u16,
 }

 impl Params {
@@ -113,6 +121,8 @@ impl Params {
             virtio_guest_cid: None,
             pci_passthrough: None,
             vm_type: VmType::Default,
+            hats_launcher_port: 8889,
+            host_forwarding_port: 50051,
         }
     }
 }
@@ -233,6 +243,17 @@ impl Qemu {
                 "hostfwd=tcp:{host_address}:{host_proxy_port}-{vm_address}:{vm_port}"
             ));
         };
+
+        // Route traffic from guest to kv-server.
+        netdev_rules.push(format!(
+        "hostfwd=tcp:{host_address}:{}-{vm_address}:50051",
+        params.host_forwarding_port
+        ));
+
+        // Route traffic from host to HATs launcher service.
+        netdev_rules.push(
+            format!("guestfwd=tcp:10.0.2.100:8889-cmd:nc {host_address} {}", params.hats_launcher_port));
+
         cmd.args(["-netdev", netdev_rules.join(",").as_str()]);
         cmd.args([
             "-device",
@@ -266,7 +287,7 @@ impl Qemu {
             "brd.rd_nr=1".to_string(),
             format!("brd.rd_size={ramdrive_size}"),
             "brd.max_part=1".to_string(),
-            format!("ip={vm_address}:::255.255.255.0::eth0:off"),
+            format!("ip={vm_address}:::255.255.255.0::enp0s1:off"),
             "quiet".to_string(),
             "--".to_string(),
             format!("--launcher-addr=vsock://{VMADDR_CID_HOST}:{launcher_service_port}"),
diff --git a/oak_containers_orchestrator/src/crypto.rs b/oak_containers_orchestrator/src/crypto.rs
index 47fac04a40..ff28f536da 100644
--- oak_containers_orchestrator/src/crypto.rs
+++ oak_containers_orchestrator/src/crypto.rs
@@ -42,7 +42,7 @@ pub fn generate_instance_keys() -> (InstanceKeys, InstancePublicKeys) {

 pub struct InstanceKeys {
     encryption_key: EncryptionKey,
-    signing_key: p256::ecdsa::SigningKey,
+    pub signing_key: p256::ecdsa::SigningKey,
 }

 pub struct InstancePublicKeys {
